### PR TITLE
resourcequota: use dynamic informer

### LINKED ISSUE #53777: quota support for custom resources
**Is this a BUG REPORT or FEATURE REQUEST?**:
FEATURE REQUEST

/kind feature

For standard resources like (pods, configmaps), we can use quota to limit the amount of these resources in cluster. However, for a user defined custom resource, we don't support quota.

### PR DESCRIPTION
**Pull request by zhouhaibing089**
The resource quota controller should use a dynamic informer so it
can create informer for custom resources.

fixes #53777

A follow up on https://github.com/kubernetes/kubernetes/pull/64201.

```release-note
Object count quota is now supported for namespaced custom resources using the `count/<resource>.<group>` syntax.
```

### CODE REVIEW

**Comment by yue9944882** [comment]
/hold
/cc @deads2k @p0lyn0mial @nikhita 

not clear whether we still want to continue/rebase on #64201

File: staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer.go [hunk 1]
```diff
@@ -62,20 +62,20 @@ type dynamicSharedInformerFactory struct {
...
 	if exists {
-		return informer
+		return informer, nil
 	}
 
 	informer = NewFilteredDynamicInformer(f.client, gvr, f.namespace, f.defaultResync, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc}, nil)
 	f.informers[key] = informer
 
-	return informer
+	return informer, nil
```
Comment (krmayankk) [inline staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer.go]: why do we need this change ?

File: pkg/controller/resourcequota/resource_quota_controller.go [hunk 2]
```diff
@@ -56,6 +57,33 @@ type InformerFactory interface {
 	Start(stopCh <-chan struct{})
 }
 
+type informerFactory struct {
```
Comment (liggitt) [inline pkg/controller/resourcequota/resource_quota_controller.go]: this doesn't seem quota-specific, and the next thing we are likely to do with it is to collapse dynamic informer building in garbage collection onto it. can we place this somewhere both GC and quota can use it?

**Comment by liggitt** [comment]
can we add an e2e test for applying quota to a CRD?

File: pkg/controller/resourcequota/resource_quota_controller.go (hunk 2 above)
Comment (zhouhaibing089) [reply pkg/controller/resourcequota/resource_quota_controller.go]: Yeah, sounds good to me, I will move it upp.

**Comment by liggitt** [comment]
ping, any plans to revisit this soon?

File: pkg/controller/resourcequota/resource_quota_controller.go (hunk 2 above)
Comment (zhouhaibing089) [reply pkg/controller/resourcequota/resource_quota_controller.go]: Moved to `pkg/controller/informer_factory.go`.

**Comment by zhouhaibing089** [comment]
I am writing the e2e test now.

**Comment by zhouhaibing089** [comment]
e2e test is also added.

**Comment by zhouhaibing089** [comment]
Interesting, the e2e test did not pass, while I can verify with a local environment... I'll look into more on the case itself.

**Comment by yue9944882** [comment]
/hold cancel

guess the previous thread has leaked

**Comment by zhouhaibing089** [comment]
```
• [SLOW TEST:278.137 seconds]
[sig-api-machinery] ResourceQuota
/Users/haibzhou/gopath/src/k8s.io/kubernetes/test/e2e/apimachinery/framework.go:22
  should create a ResourceQuota and capture the life of a custom resource.
  /Users/haibzhou/gopath/src/k8s.io/kubernetes/test/e2e/apimachinery/resource_quota.go:493
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSFeb 28 09:49:02.005: INFO: Running AfterSuite actions on all nodes
Feb 28 09:49:02.005: INFO: Running AfterSuite actions on node 1

Ran 1 of 3256 Specs in 278.185 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 3255 Skipped PASS
```

By testing it locally, looks like it takes 278s...

**Comment by zhouhaibing089** [comment]
Ping @liggitt for review.

File: test/e2e/apimachinery/resource_quota.go [hunk 3]
```diff
@@ -487,6 +488,61 @@ var _ = SIGDescribe("ResourceQuota", func() {
 		Expect(err).NotTo(HaveOccurred())
 	})
 
+	It("should create a ResourceQuota and capture the life of a custom resource.", func() {
+		By("Creating a Custom Resource Definition")
+		testcrd, err := framework.CreateTestCRD(f)
+		Expect(err).NotTo(HaveOccurred())
+		defer testcrd.CleanUp()
+		countResourceName := "count/" + testcrd.Crd.Spec.Names.Plural + "." + testcrd.Crd.Spec.Group
+		time.Sleep(45 * time.Second)
```
Comment (liggitt) [inline test/e2e/apimachinery/resource_quota.go]: this looks suspect... what is this waiting for?

Comment (zhouhaibing089) [reply test/e2e/apimachinery/resource_quota.go]: Right after the CRD creation, the resource quota controller takes time to discover this new resource via discovery, and it is 30 seconds hardcoded here: https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/core.go#L322.

If we do not wait for it here, let's say we create the quota immediately, now the resourcequota controller won't treat it as quotable yet, and treat it as synced successfully, no unless someone else trigger an update on the ResourceQuota object, it won't be queued anymore.

Though there is another loop which runs `enqueueAll` every 5 minutes to pick it up.

Comment (zhouhaibing089) [reply test/e2e/apimachinery/resource_quota.go]: ping @liggitt do you have suggestion on this?

Comment (liggitt) [reply test/e2e/apimachinery/resource_quota.go]: I'd suggest creating a test resourcequota object for the custom resource, and poll the quota object until the quota controller updates status (which indicates it discovered/resynced). Polling on a ~1 second interval means we'll continue the test as soon as possible, and giving it a longer timeout (at least 2x the quota controller sync period) means we'll be more robust to flakes.

Comment (zhouhaibing089) [reply test/e2e/apimachinery/resource_quota.go]: > I'd suggest creating a test resourcequota object for the custom resource, and poll the quota object until the quota controller updates status (which indicates it discovered/resynced)

@liggitt: This isn't going to work, as the question remains: when do we create the test resourcequota? it it is too fast, then the resource quota controller is gonna to skip it, and we end up with waiting for 5 minutes.

Comment (zhouhaibing089) [reply test/e2e/apimachinery/resource_quota.go]: oh, maybe you mean creating one test resourcequota object every second.

Comment (liggitt) [reply test/e2e/apimachinery/resource_quota.go]: we can create one, and keep updating it (set an annotation to an incrementing number or something) to keep it hitting the controller

File: test/e2e/apimachinery/resource_quota.go [hunk 4]
```diff
@@ -487,6 +488,61 @@ var _ = SIGDescribe("ResourceQuota", func() {
...
+			},
+		}, resourceClient, testcrd.Crd)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Ensuring resource quota status captures custom resource creation")
+		usedResources = v1.ResourceList{}
+		usedResources[v1.ResourceName(countResourceName)] = resource.MustParse("1")
+		err = waitForResourceQuota(f.ClientSet, f.Namespace.Name, quotaName, usedResources)
+		Expect(err).NotTo(HaveOccurred())
+
```
Comment (yliaog) [inline test/e2e/apimachinery/resource_quota.go]: could you add a test case that 'not allow creating more CRs that exceeds the remaining quota'

**Comment by yliaog** [comment]
also please add a bit more info to the release note, something like below:

Object count quota is now supported for namespaced custom resources using the `count/<resource>.<group>` syntax.

File: test/e2e/apimachinery/resource_quota.go (hunk 3 above)
Comment (zhouhaibing089) [reply test/e2e/apimachinery/resource_quota.go]: That's a good idea. 👍

Comment (zhouhaibing089) [reply test/e2e/apimachinery/resource_quota.go]: This is updated.

File: test/e2e/apimachinery/resource_quota.go (hunk 4 above)
Comment (zhouhaibing089) [reply test/e2e/apimachinery/resource_quota.go]: Yeah, sure, this is also added.

File: test/e2e/apimachinery/resource_quota.go (hunk 3 above)
Comment (zhouhaibing089) [reply test/e2e/apimachinery/resource_quota.go]: Well, I can not update annotations, as resourcequota controller only enqueue the object when `spec.Hard` differs. I will switch to update `spec.Hard` instead.

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/pull/74747 will need to merge first, to ensure we don't reintroduce deadlock possibilities with this change.

e2e test also looks like it is still failing

**Comment by liggitt** [comment]
#74747 is merged, this is good to proceed

can you rebase and take a look at the failing e2e test?

/hold cancel

**Comment by zhouhaibing089** [comment]
Rebased and looking the e2e failure now.

File: test/e2e/apimachinery/resource_quota.go [hunk 5]
```diff
@@ -1524,3 +1609,33 @@ func waitForResourceQuota(c clientset.Interface, ns, quotaName string, used v1.R
...
+			return false, nil
+		}
+		// verify that the quota shows the expected used resource values
+		_, ok := resourceQuota.Status.Used[resourceName]
+		if ok {
+			return true, nil
+		}
+
+		current := resourceQuota.Spec.Hard[resourceName]
+		current.Add(resource.MustParse("1"))
```
Comment (liggitt) [inline test/e2e/apimachinery/resource_quota.go]: this need to set something that varies each time... setting the same value will no-op in etcd and result in no watch events being delivered

Comment (zhouhaibing089) [reply test/e2e/apimachinery/resource_quota.go]: Doesn't it add `1` every time?

Comment (liggitt) [reply test/e2e/apimachinery/resource_quota.go]: ah, yes. I missed the `Add`... I thought it was setting to "1" every time

**Comment by zhouhaibing089** [comment]
I reproduce the failure locally, and the quota is not updated as I expect..

File: test/e2e/apimachinery/resource_quota.go [hunk 6]
```diff
@@ -1524,3 +1609,33 @@ func waitForResourceQuota(c clientset.Interface, ns, quotaName string, used v1.R
...
+// updateResourceQuotaUntilUsageAppears updates the resource quota object until the usage is populated
+// for the specific resource name.
+func updateResourceQuotaUntilUsageAppears(c clientset.Interface, ns, quotaName string, resourceName v1.ResourceName) error {
+	return wait.Poll(framework.Poll, 1*time.Minute, func() (bool, error) {
+		resourceQuota, err := c.CoreV1().ResourceQuotas(ns).Get(quotaName, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		// used may not yet be calculated
+		if resourceQuota.Status.Used == nil {
```
Comment (zhouhaibing089) [inline test/e2e/apimachinery/resource_quota.go]: Well, here is the bug, when `status.used` is nil, we should keep updating `spec.hard`.

**Comment by zhouhaibing089** [comment]
All fixed.

**Comment by zhouhaibing089** [comment]
Looks like all the tests are green now.

**Comment by liggitt** [comment]
/lgtm
/approve

Thanks for all the work on this

**Pull request merged**